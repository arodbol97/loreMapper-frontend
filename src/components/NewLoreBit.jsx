import React, { useState, useEffect } from 'react';
import axios from 'axios';

const NewLoreBit = ({ worldId, setShowNewForm, getLoreBits }) => {
    const divStyles = {
        display: 'flex',
        flexWrap: 'wrap',
        width: '100%',
        alignItems: 'flex-start'
    };

    const formStyles = {
        width: '100%',
    };

    const labelStyles = {
        color: 'white',
        width: '20%',
        textAlign: 'right',
        paddingRight: '10px',
        marginBottom: '5px'
    };

    const inputStyles = {
        width: 'calc(100% - 0px)',
        marginBottom: '10px',
        boxSizing: 'border-box',
        backgroundColor: 'lightgrey'
    };

    const buttonStyles = {
        width: 'calc(50% - 0px)',
        marginTop: '15px',
        boxSizing: 'border-box',
        color: 'white',
        cursor: 'pointer'
    };

    const titleStyles = {
        border: '5px solid #282c34',
        width: '100%',
        height: '30px',
        color: 'white',
        margin: '10px 0 20px 0',
        textAlign: 'center',
        backgroundColor: '#007bff'
    };

    const errorStyles = {
        width: '100%',
        textAlign: 'center',
        color: 'red',
        fontSize: '15px',
    };

    const [name, setName] = useState('');
    const [description, setDescription] = useState('');    

    const cancelForm = () => {        
        getLoreBits();
        setShowNewForm(false);
    };

    const handleChange = (event) => {
        const { name, value } = event.target;
        if (name === 'name') {
            setName(value);
        } else if (name === 'description') {
            setDescription(value);
        }
    };

    const handleSubmit = async (event) => {
        event.preventDefault();

        const formData = {
            loreBitWorld: worldId,
            loreBitTitle: name,
            loreBitDescription: description,            
        };        
        try {
            const response = await axios.post('https://loremapper-backend-b042c39916b5.herokuapp.com/loreBit/create', formData);
            const loreBitData = response.data.loreBitData;
            if (!loreBitData.created) {
                document.getElementById('formError').innerHTML = loreBitData.error;
            } else {
                cancelForm();
            }
        } catch (error) {
            console.error('Create error: ', error);
            document.getElementById('formError').innerHTML = error.message;
        }
    };

    useEffect(() => {
        return () => {
            cancelForm();
        };
    }, []);

    return (
        <div style={divStyles}>
            <h3 style={titleStyles}>Nuevo lore</h3>
            <form onSubmit={handleSubmit} style={formStyles} id='loreForm'>
                <label style={labelStyles}>Título:</label>
                <input style={inputStyles} type="text" name="name" value={name} maxLength={50} onChange={handleChange} />
                <label style={labelStyles}>Descripción:</label>
                <textarea style={{ ...inputStyles, resize: 'vertical' }} name="description" onChange={handleChange} />
                
                <button style={{ ...buttonStyles, backgroundColor: 'red', color: 'white', fontWeight: 'bold' }} type="button" onClick={cancelForm}>Cancelar</button>
                <button style={{ ...buttonStyles, backgroundColor: '#007bff', color: 'white', fontWeight: 'bold' }} type="submit">Confirmar</button>
                <span id="formError" style={errorStyles}></span>
            </form>
        </div>
    );
};

export default NewLoreBit;
