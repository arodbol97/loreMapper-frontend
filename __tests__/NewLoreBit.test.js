import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import axios from 'axios';
import NewLoreBit from '../src/components/NewLoreBit';

jest.mock('axios');


describe('NewLoreBit', () => {

    const WORLD_ID = 1;        
    const POST_RESPONSE = {
        "data": {
            "loreBitData": {
                "created": true
            }
        }
    }

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should call axios with the correct data when form is submitted', async () => {
        const getLoreBits = jest.fn();
        const setNewLoreBit = jest.fn();        
        axios.post.mockResolvedValue(POST_RESPONSE);
        const { container} = render(<NewLoreBit worldId={WORLD_ID} setShowNewForm={setNewLoreBit} getLoreBits={getLoreBits}/>);
        const nameInput = container.querySelector('#loreBitTitleForm');
        const descriptionInput = container.querySelector('#loreBitDescriptionForm');        
        const submitButton = container.querySelector('#submitLoreBitButton');
        const URL = 'https://loremapper-backend-b042c39916b5.herokuapp.com/loreBit/create'
        const EXPECTED_REQUEST_DATA = {
            loreBitWorld: WORLD_ID,  
            loreBitTitle: 'LoreBit Title',            
            loreBitDescription: "LoreBit Description",            
        };
    
        fireEvent.change(nameInput, { target: { value: 'LoreBit Title' } });
        fireEvent.change(descriptionInput, { target: { value: 'LoreBit Description' } });
        fireEvent.click(submitButton);
    
        await waitFor(() => {
          expect(axios.post).toHaveBeenCalledTimes(1);
          expect(axios.post).toHaveBeenCalledWith(URL, EXPECTED_REQUEST_DATA);
        });
      });



      it('should match snapshot', async () => {
        const getLoreBits = jest.fn();
        const setNewLoreBit = jest.fn();  
        const { asFragment } = render(<NewLoreBit worldId={WORLD_ID} setShowNewForm={setNewLoreBit} getLoreBits={getLoreBits}/>);
        expect(asFragment()).toMatchSnapshot();
      });

});
