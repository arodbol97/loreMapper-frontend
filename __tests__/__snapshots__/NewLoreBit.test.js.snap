// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewLoreBit should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-wrap: wrap; width: 100%; align-items: flex-start;"
  >
    <h3
      style="border: 5px solid #282c34; width: 100%; height: 30px; color: white; margin: 10px 0px 20px 0px; text-align: center; background-color: rgb(0, 123, 255);"
    >
      Nuevo lore
    </h3>
    <form
      id="loreForm"
      style="width: 100%;"
    >
      <label
        style="color: white; width: 20%; text-align: right; padding-right: 10px; margin-bottom: 5px;"
      >
        Título:
      </label>
      <input
        id="loreBitTitleForm"
        maxlength="100"
        name="name"
        style="width: calc(100% - 0px); margin-bottom: 10px; box-sizing: border-box; background-color: lightgrey;"
        type="text"
        value=""
      />
      <label
        style="color: white; width: 20%; text-align: right; padding-right: 10px; margin-bottom: 5px;"
      >
        Descripción:
      </label>
      <textarea
        id="loreBitDescriptionForm"
        maxlength="1000"
        name="description"
        style="width: calc(100% - 0px); margin-bottom: 10px; box-sizing: border-box; background-color: lightgrey; resize: vertical;"
      />
      <button
        style="width: calc(50% - 0px); margin-top: 15px; box-sizing: border-box; color: white; cursor: pointer; background-color: red; font-weight: bold;"
        type="button"
      >
        Cancelar
      </button>
      <button
        id="submitLoreBitButton"
        style="width: calc(50% - 0px); margin-top: 15px; box-sizing: border-box; color: white; cursor: pointer; background-color: rgb(0, 123, 255); font-weight: bold;"
        type="submit"
      >
        Confirmar
      </button>
      <span
        id="formError"
        style="width: 100%; text-align: center; color: red; font-size: 15px;"
      />
    </form>
  </div>
</DocumentFragment>
`;
